File: cynq/examples/test_full.cpp
#include "cynq/cynq.h"
#include "udmalib/udma.h"

#define max32    (0xffffffff)
#define top32(x) (x >> 32 & max32)
#define bot32(x) (x & max32)

int main(int argc, char **argv) {
  
  UdmaRepo repo;
  UdmaDevice *device = repo.device(0);
  char *udmabuf = device->map();
  
  int count = 32;
  int size = count*4;

  uint32_t *buffer0 = (uint32_t*) udmabuf;
  uint32_t *buffer1 = (uint32_t*) &buffer0[count];
  uint32_t *buffer2 = (uint32_t*) &buffer0[2*count];
  
  long buf0addr = device->phys_addr;
  long buf1addr = device->phys_addr + size;
  long buf2addr = device->phys_addr + 2*size;
  
  for (int i = 0; i < count; i++) {
    buffer0[i] = i;
    buffer1[i] = i;
    buffer2[i] = -1;
  }


  PRManager prmanager;

  //StaticAccelInst vecadd = prmanager.fpgaLoadStatic("Full_vecadd");

  prmanager.fpgaLoadShell("Ultra96_100MHz_2");
  AccelInst vecadd = prmanager.fpgaLoad("Partial_vadd");

  paramlist params({
    {"group_id_x", 0},
    {"group_id_y", 0},
    {"group_id_z", 0},
    {"global_offset_x", 0},
    {"global_offset_y", 0},
    {"global_offset_z", 0},
    {"ina_1", bot32(buf0addr)},
    {"ina_2", top32(buf0addr)},
    {"inb_1", bot32(buf1addr)},
    {"inb_2", top32(buf1addr)},
    {"out_r_1", bot32(buf2addr)},
    {"out_r_2", top32(buf2addr)},
  });


  vecadd.programAccel(params);
  vecadd.runAccel();
  vecadd.wait();
  
  for (int i = 0; i < count; i++) {
    std::cout << i << ": " << buffer2[i] << std::endl;
  }
 
  
}

File: ponq/examples/test_full.py
#!/usr/bin/env python3
import sys
sys.path.append("../..")
sys.path.append("..")

import udmalib.udma as udma
from ponq import Ponq
import cv2, numpy as np

# bit hacking functions for driving unit
fixpfactor = 1 << 29
int32max = (1 << 32) - 1
def float2fix(value): return int(value * fixpfactor)
def top32(data): return int((data >> 32) & int32max)
def bot32(data): return int(data & int32max)

# load udma buffer for hardware use
udmarepo = udma.udmas()
udmadev = udmarepo.getDevice(0)
ubuf = udmadev.map()
  
# hardware module parameters
udma_addy = udmadev.phys_addr

items = 1024
size = items*4

buf0addr = udmadev.phys_addr
buf1addr = udmadev.phys_addr + size
buf2addr = udmadev.phys_addr + 2*size


for i in range(items):
  byte_val = int(i).to_bytes(4, byteorder=sys.byteorder)
  null_val = int(1).to_bytes(4, byteorder=sys.byteorder)
  for b in range(4):
    ubuf[0*size + i*4 + b] = byte_val[b]
    ubuf[1*size + i*4 + b] = byte_val[b]
    ubuf[2*size + i*4 + b] = null_val[b]

# initialise ponq
runs = items // 16

manager = Ponq("../../bitstreams", "../../build/bit_patch_bin")
acc0 = manager.load("Full_vadd")
acc0.writeReg("group_id_x", 0)
acc0.writeReg("group_id_y", 0)
acc0.writeReg("group_id_z", 0)
acc0.writeReg("global_offset_x", 0)
acc0.writeReg("global_offset_y", 0)
acc0.writeReg("global_offset_z", 0)
acc0.writeReg("ina_1", bot32(buf0addr))
acc0.writeReg("ina_2", top32(buf0addr))
acc0.writeReg("inb_1", bot32(buf1addr))
acc0.writeReg("inb_2", top32(buf1addr))
acc0.writeReg("out_r_1", bot32(buf2addr))
acc0.writeReg("out_r_2", top32(buf2addr))

for run in range(runs-1):
  acc0.writeReg("group_id_x", run)
  acc0.run()
  acc0.wait()

acc0.unload()

for i in range(items):
  byte_val = ubuf[2*size + i*4:2*size + (i+1)*4]
  int_val = int.from_bytes(byte_val, byteorder=sys.byteorder)
  print(str(i) + ": " + str(int_val))


